File: Algorithm/ReformatPhoneNumber.java
// Solution:
// Time complexity: O(n)
// Space complexity: O(1)
class Solution {
    public String reformatNumber(String number) {
        StringBuilder ans = new StringBuilder();
        int total = 0;
        for (char c : number.toCharArray())
            if (Character.isDigit(c)) ++total; // 这里if执行条件可以不用大括号包住。
        int l = 0;
        for (char c : number.toCharArray()) {
            if (!Character.isDigit(c)) continue;
            ans.append(c);
            ++l;
            if ((l % 3 == 0 && total - l >= 4) || 
                (total % 3 != 0 && total - l == 2) || 
                (total % 3 == 0 && total - l == 3))
                ans.append('-');
        }
        return ans.toString();
    }
}

File: Algorithm/ReformatPhoneNumber.cpp
class Solution {
public:
    string reformatNumber(string number) {
        string ans;
        const int total = count_if(begin(number), end(number),
                                  [](char c) { return isdigit(c); });
        int l = 0;
        for (char c : number) {
            if (!isdigit(c)) continue;
            ans += c;
            ++l;
            if ((l % 3 == 0 && total - l >= 4) ||
                (total % 3 != 0 && total - l == 2)
                || (total % 3 == 0 && total - l == 3))
             ans += "-";
        }
        return ans;
    }
};
